go/mcap/indexed_message_iterator.go

```diff
@@ -286,28 +432,85 @@ func (it *indexedMessageIterator) Next(_ []byte) (*Schema, *Channel, *Message, e
 		}
 	}
 
-	for it.indexHeap.Len() > 0 {
-		ri, err := it.indexHeap.HeapPop()
-		if err != nil {
-			return nil, nil, nil, err
-		}
-		if ri.messageIndexEntry == nil {
-			err := it.loadChunk(ri.chunkIndex)
-			if err != nil {
+	if !it.useHeap {
+		if it.curMessageIndex >= len(it.messageIndexes) {
+			if it.curChunkIndex >= len(it.chunkIndexes) {
+				return nil, nil, nil, io.EOF
+			}
+			chunkIndex := it.chunkIndexes[it.curChunkIndex]
+			if err := it.loadChunk(chunkIndex); err != nil {
 				return nil, nil, nil, err
 			}
-			continue
+			it.curMessageIndex = 0
+			it.curChunkIndex++
+			return it.Next2(msg)
 		}
-		chunkOffset := ri.messageIndexEntry.Offset
-		length := binary.LittleEndian.Uint64(ri.buf[chunkOffset+1:])
-		messageData := ri.buf[chunkOffset+1+8 : chunkOffset+1+8+length]
-		message, err := ParseMessage(messageData)
-		if err != nil {
+		messageIndex := it.messageIndexes[it.curMessageIndex]
+		decompressedChunk := &it.decompressedChunks[0]
+		if err := loadMessageAtOffset(decompressedChunk.buf, messageIndex.Offset, msg); err != nil {
 			return nil, nil, nil, err
 		}
-		channel := it.channels[message.ChannelID]
-		schema := it.schemas[channel.SchemaID]
-		return schema, channel, message, nil
+		decompressedChunk.unreadMessages--
+		it.curMessageIndex++
+		schema, channel, err := it.getSchemaAndChannel(msg.ChannelID)
+		return schema, channel, msg, err
```

Comment:
this would be a little clearer for me to read with an explicit error check:

```
schema, channel, err := it.getSchemaAndChannel(msg.ChannelID)
if err != nil {
  return nil, nil, nil, err
}
return schema, channel, msg, nil
```

that tells the reader that for sure this library won't return real values along with a non-nil error.

The other reason IMO to always do this and avoid the shortcut, is you will sometimes find libraries that do stuff like this:

```
package main

import "fmt"

type MyError struct{}

func (e MyError) Error() string {
	return fmt.Sprintf("uh oh")
}

func foo() *MyError {
	return nil
}

func bar() error {
	err := foo()
	return err
}

func main() {
	err := bar()
	if err != nil {
		fmt.Printf("error! %+v", err)
		return
	}
}
```

i.e they will type the return value with their custom error type, rather than Error. In this example bar() implements the shortcut approach. When you run this program you will get a surprise:

```
[~] $ go run main.go 
error! <nil>
```

As a library consumer you usually won't pay attention to how the library deals with errors until you have a need to. So better to always check the errors and be explicit. This program doesn't have the problem:

```
package main

import "fmt"

type MyError struct{}

func (e MyError) Error() string {
	return fmt.Sprintf("uh oh")
}

func foo() *MyError {
	return nil
}

func bar() error {
	err := foo()
	if err != nil {
		return err
	}
	return nil
}

func main() {
	err := bar()
	if err != nil {
		fmt.Printf("error! %+v", err)
		return
	}
}
```